Feature 941

anushka-singh opened the pull request:
### Feature or Bugfix
- Feature

### Detail
Share requests running in parallel override one another.

### Relates
- https://github.com/data-dot-all/dataall/issues/941

### Testing
- 2 simultaneous shares got processed successfully with locking mechanism.
- Here is what the dataset_lock DB looks like and the values in it.
<img width="1386" alt="Screenshot 2024-02-14 at 5 12 54 PM" src="https://github.com/data-dot-all/dataall/assets/26413731/9e36044f-4f3d-4371-a6f2-7dee399da7a7">
- acquiredBy column show the share which last acquired the lock for the particular dataset


### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

anushka-singh commented on backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:
```diff
@@ -226,3 +246,108 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+
+    @staticmethod
+    def acquire_lock_with_retry(dataset_uri, session, share_uri):
+        max_retries = 10
+        retry_interval = 60
```
These 2 numbers are open to discussion. 
@dlpzx @noah-paige What do you think of these numbers?

noah-paige replied:
I think these numbers make sense - will do some testing to get a better sense of user experience but think 10 minute max wait seems good

Should we make these global variables at the top of this file, such as 
```
MAX_RETRIES = 10
RETRY_INTERVAL = 60
```

anushka-singh replied:
Yes I will make them global variables. 

We could also make the max_retries 20 and keep the retry_interval 30. That would still give us 10 min coverage but only a 30 sec delay between retries. However, this could cause heavy throttling.

anushka-singh commented:
Filed another issue for opening a discussion on another locking usecase: https://github.com/data-dot-all/dataall/issues/1074.

noah-paige commented on backend/dataall/modules/datasets_base/db/dataset_models.py:
```diff
@@ -164,3 +164,14 @@ class DatasetBucket(Resource, Base):
+
+
+class DatasetLock(Base):
+    __tablename__ = 'dataset_lock'
+    datasetUri = Column(String, nullable=False, primary_key=True)
+    isLocked = Column(Boolean, default=False)
+    acquiredBy = Column(String, nullable=False)
+
```
The one thing that comes to mind is maybe we should have a foreign key relation between this table and datasetUri in dataset table?

anushka-singh replied:
Sure, I can create that.

anushka-singh commented on backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:
```diff
@@ -38,71 +43,86 @@ def approve_share(cls, engine: Engine, share_uri: str) -> bool:
+                    target_environment,
+                ) = ShareObjectRepository.get_share_data(session, share_uri)
+
+                lock_acquired = cls.acquire_lock_with_retry(dataset.datasetUri, session, share.shareUri)
+                if not lock_acquired:
+                    log.error(f"Failed to acquire lock for dataset {dataset.datasetUri}. Exiting...")
+                    return False
```
Currently, if a share is unable to acquire a lock, it remains in share_approved state. It does not go into share_failed - which is problematic since the user will not be able to revoke or delete the share to recreate it later. 
Although its an extremely rare situation to occur that within 10 mins multiple shares are approved and are unable to acquire locks, we must add a safeguard in place. 

I wanted to hear your thoughts on how to achieve this here.
I could create a new SM and use transition states to transition to share_failed. Not sure if thats the best approach though.

noah-paige replied:
When share is approved...
- Share Object status goes from `Submitted` --> `Approved`
- Share Item(s) statuses go from `Pending_Approval` --> `Share_Approved`

I think that would be best as you proposed, have a new Transition that would bring 
- Share Object status from `Approved` --> `Processed`
- Each Share Item(s) statuses from `Share_Approved` --> `Share_Failed`
if the timeout is reached and can not acquire the lock

In the case above then we should not have any share items or share object stuck in `Share_Approved` or `Approved` , respectively and the use than retry to process the share items at a later time

anushka-singh replied:
Done!

anushka-singh commented:
Alembic script failing with :
```
ERROR [alembic.util.messaging] Multiple head revisions are present for given argument 'head'; please specify a specific target revision, '<branchname>@head' to narrow to a specific head, or 'heads' for all heads
FAILED: Multiple head revisions are present for given argument 'head'; please specify a specific target revision, '<branchname>@head' to narrow to a specific head, or 'heads' for all heads
```

I am not sure how to specify which head version alembic should run upgrade with in the script. @noah-paige Can you suggest?

noah-paige commented:
> Alembic script failing with :
> 
> ```
> ERROR [alembic.util.messaging] Multiple head revisions are present for given argument 'head'; please specify a specific target revision, '<branchname>@head' to narrow to a specific head, or 'heads' for all heads
> FAILED: Multiple head revisions are present for given argument 'head'; please specify a specific target revision, '<branchname>@head' to narrow to a specific head, or 'heads' for all heads
> ```
> 
> I am not sure how to specify which head version alembic should run upgrade with in the script. @noah-paige Can you suggest?

@anushka-singh - I believe it is because there are 2 alembic migration scripts building off of the same `down_revision`

A PR we merged last week has revision id of `6c9a8afee4e4` which should be the latest in OS and the migration script in this PR should be updated to have that revision id as its down revision

noah-paige commented on backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:
```diff
@@ -226,3 +249,106 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+            query.update(
+                {
+                    "isLocked": False,
+                    "acquiredBy": share_uri
```
do we need to update  `"acquiredBy": share_uri` - is it not already this share_uri and we can just update `isLocked`. If anything I think it makes most sense to update `"acquiredBy": ''` if the lock is being released (only if `acquiredBy` can be nullable

anushka-singh replied:
The acquiredBy column was more to indicate lock last acquired by which share. 
However, I can make it nullable and reset it to '' when we release lock too. 
I just thought knowing the last share that acquired it, might be useful in debugging. 

I have pushed a revision where acquiredBy can be null and acquiredBy = '' while releasing lock.

anushka-singh commented on backend/migrations/versions/a1049d854e29_implement_dataset_locking_mechanism.py:
```diff
@@ -0,0 +1,152 @@
+                dataset_lock = DatasetLock(
+                    datasetUri=dataset.datasetUri,
+                    isLocked=False,
+                    acquiredBy=''
```
Sure! Made the change

anushka-singh commented on backend/dataall/modules/dataset_sharing/db/share_object_repositories.py:
```diff
@@ -1307,3 +1307,61 @@ def count_role_principal_shares(session, principal_id: str, principal_type: Prin
         )
+
+
+class DatasetLockShareItemManagementSM:
```
I thought it best to keep it different from ShareItemSM and ShareObjectSM since the transformations in those are very different from the ones we wanted. For example, we are going directly from Approved to Processed - which might confuse some customers.

noah-paige replied:
I think we should opt to have 1 SM manage the entire status lifecycle of the share object / share items rather than 2 separate SM classes that would manage different transitions of the same status attribute

noah-paige replied:
I think best to just add a new `ShareObjectActions` of `FailedStart` or `FailedAcquireLock` or similar name and have that start the transitions we need using the pre-existing SM classes

anushka-singh replied:
Sure, I have included the change in the next revision.

TejasRGitHub commented on backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:
```diff
@@ -226,3 +277,127 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+            session.commit()
+            return True
+        except Exception as e:
+            log.error("Error occurred while releasing lock:", e)
```
Should there be a similar session.rollback() in case anything fails ?

anushka-singh replied:
Added!

TejasRGitHub commented on backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:
```diff
@@ -226,3 +277,127 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+        try:
+            # Execute the query to get the DatasetLock object
+            dataset_lock = (
+                session.query(DatasetLock)
```
Can there be a situation in which two ECS processes start at the same time and arrive here and read the same state . And then update the dataset_lock table on L308 - 311. 

Similar to how it is mentioned here - https://dev.to/ivankwongtszfung/safe-update-operation-in-postgresql-using-sqlalchemy-3ela

noah-paige replied:
Reading the above docs it seems like this is true - should we include `.with_for_update().first()` as specified in the docs @TejasRGitHub posted cc: @anushka-singh 

I believe that will resolve as no other session can update that record and will return an OperationalError that we can handle in exceptions

anushka-singh replied:
@TejasRGitHub, In my code, I had added .first() to fetch only the first entry. So, I dont think a process will be able to read 2 states at the same time. 

I have anyway added the `.with_for_update().first()` guardrail to be safe. Thanks for the suggestion!

noah-paige commented on backend/dataall/modules/dataset_sharing/services/data_sharing_service.py:
```diff
@@ -226,3 +275,124 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+                    )
+                )
+                .with_for_update().one()
+            )
```
@anushka-singh - sorry for all the back and forth ...

with the latest changes to add `.with_for_update().one()` I am now receiving error: `"Arguments: (AttributeError("'DatasetLock' object has no attribute 'update'"),)"`


I think for we have 2 options:

- (1) For `def release_lock()` we can handle the update the same as  for acquire lock 
```
            dataset_lock = (
                session.query(DatasetLock)
                .filter(
                    and_(
                        DatasetLock.datasetUri == dataset_uri,
                        DatasetLock.isLocked == True,
                        DatasetLock.acquiredBy == share_uri
                    )
                )
                .with_for_update().one()
            )
           if dataset_lock:
                dataset_lock.isLocked = False
                dataset_lock.acquiredBy = ''
...
```

- (2) OR perform query and update all in 1:
```
            query = (
                session.query(DatasetLock)
                .filter(
                    and_(
                        DatasetLock.datasetUri == dataset_uri,
                        DatasetLock.isLocked == True,
                        DatasetLock.acquiredBy == share_uri
                    )
                )
                .update(
                {
                    "isLocked": False,
                    "acquiredBy": ''
                },
                )
            )
``` (and likely can handle `acquire_lock()` similarly)

anushka-singh replied:
Hmm, that is really weird. I did not come across this issue in my testing!

But I have pushed out another revision with solution (1). Please LMK if its fixed now. I will also try it out again on my end.

noah-paige replied:
will do - trying again now

noah-paige commented:
One (..hopefully last) issue I am running into with `def acquire_lock()` (and potentially `def release_lock()`) - When running multiple shares all at once the ones that need to wait to acquire lock run into an error similar to below once beginning to process the share:
`sqlalchemy.orm.exc.DetachedInstanceError: Instance <Dataset at 0x7f8f546c4190> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3)`

I believe this is because one the first iteration when the share tries to acquire the lock and fails - since we are using the `.one()` and not `.first()` (or `.one_or_none()`) for the query, we get an error `sqlalchemy.orm.exc.NoResultFound` and thus go to the excpetion block and run `session.rollback()`

Now `session.rollback()` closes the session and makes the previous objects retrieved from the session stale and "boundless" (i.e. dataset, share object, etc. all from `get_share_data()` and `get_share_data_items()`)

To resolve we have a couple of options:
- (1) use `.first()` or similar in the query so `dataset_lock` can be `None` and we do not always go to exception block if no results found
- (2) use `session.expunge_all()` before we run `session.rollback()` (definitely required for `acquire_lock()` - maybe not for `release_lock()` as we are done processing share at this point)

I would think best if we implement both 1 and 2 above - should both be quick changes :)

anushka-singh commented:
> One (..hopefully last) issue I am running into with `def acquire_lock()` (and potentially `def release_lock()`) - When running multiple shares all at once the ones that need to wait to acquire lock run into an error similar to below once beginning to process the share: `sqlalchemy.orm.exc.DetachedInstanceError: Instance <Dataset at 0x7f8f546c4190> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3)`
> 
> I believe this is because one the first iteration when the share tries to acquire the lock and fails - since we are using the `.one()` and not `.first()` (or `.one_or_none()`) for the query, we get an error `sqlalchemy.orm.exc.NoResultFound` and thus go to the excpetion block and run `session.rollback()`
> 
> Now `session.rollback()` closes the session and makes the previous objects retrieved from the session stale and "boundless" (i.e. dataset, share object, etc. all from `get_share_data()` and `get_share_data_items()`)
> 
> To resolve we have a couple of options:
> 
> * (1) use `.first()` or similar in the query so `dataset_lock` can be `None` and we do not always go to exception block if no results found
> * (2) use `session.expunge_all()` before we run `session.rollback()` (definitely required for `acquire_lock()` - maybe not for `release_lock()` as we are done processing share at this point)
> 
> I would think best if we implement both 1 and 2 above - should both be quick changes :)
> 
> Also please let me know if you are experiencing similar

In my testing, I have not come across this issue. I am not sure why you are facing this..

I have implemented both 1 and 2 solutions. 
Implemented `session.expunge_all()` in release_lock as well to add guardrails.

noah-paige reviewed (approved):
Tested in AWS for approving and revoking 10+ shares at the same time - all works perfectly, thanks for your contribution @anushka-singh
